### PR TITLE
feat: client cache

### PR DESCRIPTION
**Pull request by willscott**
Adds a caching layer to the client

### CODE REVIEW

File: client/cache.go [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+	"time"
+
+	"github.com/drand/drand/log"
+	lru "github.com/hashicorp/golang-lru"
+)
+
+// NewCachingClient is a meta client that stores an LRU cache of
+// recently fetched random values.
+func NewCachingClient(client Client, size int, log log.Logger) (Client, error) {
+	cache, err := lru.New(size)
```
Comment (Kubuxu) [inline client/cache.go:15]: I would propose using ARC cache from this package. It solves problems with cache trashing.

Comment (willscott) [reply client/cache.go]: done

File: client/cache.go [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+		c.subscribers = c.subscribers[:0]
+
+		for _, s := range curr {
+			if s.ctx.Err() == nil {
+				c.subscribers = append(c.subscribers, s)
+				if m != nil {
+					select {
+					case s.c <- m:
+					default:
+						c.log.Warn("msg", "dropped watch message to subscriber. full channel")
```
Comment (alanshaw) [inline client/cache.go:102]: Is there a way to do this without dropping results?

Comment (willscott) [reply client/cache.go:102]: This only occurs when the client who has been given the channel isn't draining it. the channel can allow 5 results to back up internally (we could make that more), but that means nobody has pulled values off the channel for 2.5 minutes, which seems like a reasonable bound.

File: client/cache.go [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+	}
+	return &cachingClient{
+		backing: client,
+		cache:   cache,
+		log:     log,
+	}, nil
+}
+
+type cachingClient struct {
+	backing Client
```
Comment (nikkolasg) [inline client/cache.go:26]: Why not doing interface embedding, it seems more idiomatic and reduce code:
```go
type cachingClient struct {
   Client
   cache *lru.ARCCache
...
```

Comment (nikkolasg) [reply client/cache.go:26]: No need to implement watch and roundat

File: client/client.go [hunk 4]
```diff
@@ -93,6 +115,34 @@ func WithHTTPGetter(getter HTTPGetter) Option {
...
+func WithCacheSize(size int) Option {
+	return func(cfg *clientConfig) error {
+		cfg.cacheSize = size
+		return nil
+	}
+}
+
+// WithoutAggregation disables multiple calls to `Watch` being served by
+// a single underlying http poll.
+func WithoutAggregation() Option {
```
Comment (nikkolasg) [inline client/client.go:129]: Why would we ever want to disable this ?

File: client/cache.go [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+	c.subscribers = append(c.subscribers, sub)
+
+	if len(c.subscribers) == 1 {
+		ctx, cancel := context.WithCancel(context.Background())
+		go c.distribute(c.backing.Watch(ctx), cancel)
+	}
+	return sub.c
+}
+
+func (c *cachingClient) distribute(in <-chan Result, cancel context.CancelFunc) {
```
[the remark this answers is not in the document]
Comment (willscott) [reply client/cache.go]: Split 🍌

File: client/cache.go [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+		aCtx = c.subscribers[0].ctx
+		c.subscriberLock.Unlock()
+
+		var m Result
+		select {
+		case res, ok := <-in:
+			if !ok {
+				return
+			}
+			c.cache.Add(res.Round(), res)
```
[the remark this answers is not in the document]
Comment (willscott) [reply client/cache.go]: after the split, there's no caching of watch results. it doesn't seem worth the extra goroutine in the watch method

File: client/util.go [hunk 7]
```diff
@@ -69,3 +70,84 @@ func pollingWatcher(ctx context.Context, client Client, group *key.Group, log lo
...
+func (c *watchAggregator) Watch(ctx context.Context) <-chan Result {
+	c.subscriberLock.Lock()
+	defer c.subscriberLock.Unlock()
+
+	sub := subscriber{ctx, make(chan Result, 5)}
+	c.subscribers = append(c.subscribers, sub)
+
+	if len(c.subscribers) == 1 {
+		ctx, cancel := context.WithCancel(context.Background())
+		go c.distribute(c.Client.Watch(ctx), cancel)
```
Comment (nikkolasg) [inline client/util.go:106]: Am i understanding it right that this go routine may be triggered depend on the outside's requests ? If so, it doesn't seem very safe to do so without bounding the number of go routines - We'd usually want a consumer/producer model to handle things like that no ?

File: client/client.go (hunk 4 above)
Comment (willscott) [reply client/client.go:129]: 🤷‍♂️

File: client/util.go (hunk 7 above)
Comment (willscott) [reply client/util.go:106]: it'll spin up the background routine when number of subscribers goes from 0 to 1. the goroutine ends itself when subscribers goes from 1 to 0.

Comment (nikkolasg) [reply client/util.go:106]: Sorry I had my mind on the http server -_-" all good.

File: client/client.go (hunk 4 above)
Comment (nikkolasg) [reply client/client.go:129]: If we dont even know ourselves, then I would remove the option alltogether.

**Pull request merged**